<h2 style = "color:blue"> 1- Spliting Data </h2>

In [2]:
import numpy as np
import pandas as pd
data = pd.read_csv("California_Houses.csv")
#print(data.info())
N=data.shape[0]     #20640 rows
indices=np.random.permutation(N)
training=data.iloc[indices[:int(0.7*N)]]
validation=data.iloc[indices[int(0.7*N):int(0.85*N)]]
test=data.iloc[indices[int(0.85*N):]]

x_training=training.drop("Median_House_Value",axis=1).values
y_training=training["Median_House_Value"].values

x_validation=validation.drop("Median_House_Value",axis=1).values
y_validation=validation["Median_House_Value"].values

x_test=test.drop("Median_House_Value",axis=1).values
y_test=test["Median_House_Value"].values


<h2 style = "color:blue"> 2- Linear Regression from scratch </h2>

<h2 style = "color:red"> 2.1- Computing weights </h2>

<h2 style = "color:green"> Feature Scaling </h2>

In [181]:
x_mean = x_training.mean(axis=0)
x_std = x_training.std(axis=0)
x_std[x_std == 0] = 1


x_training_scaled = (x_training - x_mean)/x_std
x_validation_scaled = (x_validation - x_mean)/x_std
x_test_scaled = (x_test - x_mean)/x_std

x_training_b_scaled=np.c_[np.ones((len(x_training_scaled),1)),x_training_scaled]
x_validation_b_scaled=np.c_[np.ones((len(x_validation_scaled),1)),x_validation_scaled]
x_test_b_scaled=np.c_[np.ones((len(x_test_scaled),1)),x_test_scaled]



In [182]:
y_mean = y_training.mean()
y_std = y_training.std()

y_training_scaled = (y_training - y_mean)/y_std
y_validation_scaled = (y_validation - y_mean)/y_std
y_test_scaled = (y_test - y_mean)/y_std

In [183]:
#Training
wn= (np.linalg.inv((x_training_b_scaled.T)@x_training_b_scaled))@(x_training_b_scaled.T)@y_training_scaled
h1=x_training_b_scaled @ wn

#Validation
h2=x_validation_b_scaled @ wn

#Testing
h3 = x_test_b_scaled @ wn

<h2 style = "color:green"> MSE & MAE </h2>

In [184]:
#Training
mse1=(h1-y_training_scaled)**2
j1=0.5*np.mean(mse1)
print(f"MSE for Training = {j1}")

mae1=np.abs(h1-y_training_scaled)
j2=0.5*np.mean(mae1)
print(f"MAE for Training = {j2}")

#Validation
mse2=(h2-y_validation_scaled)**2
j3=0.5*np.mean(mse2)
print(f"MSE for Validation normal = {j3}")

mae2=np.abs(h2-y_validation_scaled)
j4=0.5*np.mean(mae2)
print(f"MAE for Validation normal = {j4}")

#Testing
mse3=(h3-y_test_scaled)**2
j5=0.5*np.mean(mse3)
print(f"MSE for Testing normal = {j5}")

mae3=np.abs(h3-y_test_scaled)
j6=0.5*np.mean(mae2)
print(f"MAE for Testing normal = {j6}")


MSE for Training = 0.17409465042637837
MAE for Training = 0.21559711519624455
MSE for Validation normal = 0.17758797247984237
MAE for Validation normal = 0.21347017532167545
MSE for Testing normal = 0.1972764279153282
MAE for Testing normal = 0.21347017532167545


<h2 style = "color:red"> 2.2- Gradient Descent </h2>

<h2 style = "color:green"> Weights & Learning Rate </h2>

In [185]:
#Training
wg = np.zeros(x_training_b_scaled.shape[1])
interations=1000
alpha=0.1
m = len(x_training_b_scaled)
for i in range(interations) :
    hg = x_training_b_scaled @ wg
    grad = (1/m)*((x_training_b_scaled.T)@(hg-y_training_scaled))
    wg = wg - alpha*grad

    if i % 100 == 0 :
        losses = 0.5 * np.mean((hg-y_training_scaled)**2)
        print(f"Iinterations = {i} , MSE = {losses} ")
print(f"MSE for Training Gradient = {0.5 * np.mean((x_training_b_scaled @ wg - y_training_scaled)**2)}")
print(f"MAE for Training Gradient = {0.5 * np.mean(np.abs(x_training_b_scaled @ wg - y_training_scaled))}")

Iinterations = 0 , MSE = 0.4999999999999999 
Iinterations = 100 , MSE = 0.18083235796839764 
Iinterations = 200 , MSE = 0.17738539542058324 
Iinterations = 300 , MSE = 0.17634084750265025 
Iinterations = 400 , MSE = 0.17585149922931065 
Iinterations = 500 , MSE = 0.17555858814622397 
Iinterations = 600 , MSE = 0.1753577990364687 
Iinterations = 700 , MSE = 0.1752096368967308 
Iinterations = 800 , MSE = 0.17509551143435134 
Iinterations = 900 , MSE = 0.1750049411071255 
MSE for Training Gradient = 0.17493126415125143
MAE for Training Gradient = 0.21662556595180915


In [186]:
#Validation
hgv = x_validation_b_scaled @ wg
mse_v=(hgv-y_validation_scaled)**2
jv1=0.5*np.mean(mse_v)
print(f"MSE for Validation Gradient = {jv1}")

mae_v=np.abs(hgv-y_validation_scaled)
jv2=0.5*np.mean(mae_v)
print(f"MAE for Validation Gradient = {jv2}")

MSE for Validation Gradient = 0.17846654262227982
MAE for Validation Gradient = 0.21467632426654154


In [187]:
#Testing
hgt = x_test_b_scaled @ wg
mse_t=(hgt-y_test_scaled)**2
jt1=0.5*np.mean(mse_t)
print(f"MSE for Testing Gradient = {jt1}")

mae_t=np.abs(hgt-y_test_scaled)
jt2=0.5*np.mean(mae_t)
print(f"MAE for Testing Gradient = {jt2}")

MSE for Testing Gradient = 0.19847876740682818
MAE for Testing Gradient = 0.2235899216093899


<h2 style = "color:blue"> 3- Linear Regression using Scikit-learn
 </h2>

In [188]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error , mean_absolute_error

model = LinearRegression()
model.fit(x_training_scaled, y_training_scaled)

w_s=model.coef_
alpha_s=model.intercept_
print(f"Weights = {w_s}")
print(f"Bias = {alpha_s}")

Weights = [ 0.64180764  0.0986099  -0.10791831  0.34279869 -0.44159082  0.23839716
 -0.81169803 -0.4526105  -0.09978255 -0.3298261   0.62798065  0.28339768
 -0.27618033]
Bias = 4.326387054313377e-15


In [192]:
#Training
y_pred1 = model.predict(x_training_scaled)
mse_s1 = mean_squared_error(y_training_scaled, y_pred1)
mae_s1 = mean_absolute_error(y_training_scaled, y_pred1)

print(f"MSE for Training Scikit Learn = {0.5 * mse_s1}")
print(f"MAE for Training Scikit Learn = {0.5*mae_s1}")

MSE for Training Scikit Learn = 0.17409465042637842
MAE for Training Scikit Learn = 0.21559711519624436


In [ ]:
#Validation
y_pred = model.predict(x_validation_scaled)
mse_s = mean_squared_error(y_validation_scaled, y_pred)
mae_s = mean_absolute_error(y_validation_scaled, y_pred)

print(f"MSE for Validation Scikit Learn = {0.5 * mse_s}")
print(f"MAE for Validation Scikit Learn = {0.5*mae_s}")

MSE for Validation Scikit Learn = 0.1775879724798424
MAE for Validation Scikit Learn = 0.2134701753216753


In [191]:
#Testing
y_pred_t = model.predict(x_test_scaled)
mse_s_t = mean_squared_error(y_test_scaled, y_pred_t)
mae_s_t = mean_absolute_error(y_test_scaled, y_pred_t)

print(f"MSE for Testing Scikit Learn = {0.5 * mse_s_t}")
print(f"MAE for Testing Scikit Learn = {0.5*mae_s_t}")

MSE for Testing Scikit Learn = 0.19727642791532796
MAE for Testing Scikit Learn = 0.22197191528752047


<h2 style = "color:purple"> 4- Finally, Comparissson & comments </h2>

<h3 style="color: #2e64fe; border-bottom: 2px solid #2e64fe; padding-bottom: 5px;"> Comparison: Normal Equation vs. Gradient Descent</h3>

<div style="background-color: #fcfcfc; padding: 15px; border-radius: 8px; border: 1px solid #e0e0e0;">


<table style="width:100%; border-collapse: collapse; margin: 10px 0;">
  <tr style="background-color: #f2f2f2;">
    <th style="padding: 10px; border: 1px solid #ddd;">Feature</th>
    <th style="padding: 10px; border: 1px solid #ddd;">Normal Equation (Closed-form)</th>
    <th style="padding: 10px; border: 1px solid #ddd;">Gradient Descent (Iterative)</th>
  </tr>
  <tr>
    <td style="padding: 10px; border: 1px solid #ddd; font-weight: bold;">Complexity</td>
    <td style="padding: 10px; border: 1px solid #ddd;">Directly computes the solution using matrix inversion.</td>
    <td style="padding: 10px; border: 1px solid #ddd;">Requires multiple iterations to converge.</td>
  </tr>
  <tr>
    <td style="padding: 10px; border: 1px solid #ddd; font-weight: bold;">Hyperparameters</td>
    <td style="padding: 10px; border: 1px solid #ddd;">None (No alpha or iterations to choose).</td>
    <td style="padding: 10px; border: 1px solid #ddd;">Needs careful tuning of Learning Rate (alpha).</td>
  </tr>
  <tr>
    <td style="padding: 10px; border: 1px solid #ddd; font-weight: bold;">Scaling</td>
    <td style="padding: 10px; border: 1px solid #ddd;">Works fine without feature scaling.</td>
    <td style="padding: 10px; border: 1px solid #ddd;">Feature scaling is <b>essential</b> for performance.</td>
  </tr>
  <tr>
    <td style="padding: 10px; border: 1px solid #ddd; font-weight: bold;">Scalability</td>
    <td style="padding: 10px; border: 1px solid #ddd;">Slow if the number of features is very large.</td>
    <td style="padding: 10px; border: 1px solid #ddd;">Scales very well with large datasets and many features.</td>
  </tr>
</table>

<p style="margin-top: 15px;">
<strong> Conclusion:</strong><br>
For this dataset, the <b>Normal Equation</b> was the most efficient as it gave the exact answer instantly. However, <b>Gradient Descent</b> is more flexible and is the method used in deep learning and larger datasets where matrix inversion is too expensive.
</p>

</div>

<h3 style="color: #2e64fe; border-bottom: 2px solid #2e64fe; padding-bottom: 5px;">📊 Models Performance Comparison</h3>

<table style="width:100%; border-collapse: collapse; margin: 20px 0; font-family: Arial, sans-serif;">
  <thead>
    <tr style="background-color: #f2f2f2; text-align: left;">
      <th style="padding: 12px; border: 1px solid #ddd;">Metric (MSE)</th>
      <th style="padding: 12px; border: 1px solid #ddd;">Normal Equation</th>
      <th style="padding: 12px; border: 1px solid #ddd;">Gradient Descent</th>
      <th style="padding: 12px; border: 1px solid #ddd;">Scikit-Learn</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="padding: 12px; border: 1px solid #ddd; font-weight: bold;">Training</td>
      <td style="padding: 12px; border: 1px solid #ddd;">0.17409</td>
      <td style="padding: 12px; border: 1px solid #ddd;">0.17493</td>
      <td style="padding: 12px; border: 1px solid #ddd;">0.17409</td>
    </tr>
    <tr style="background-color: #f9f9f9;">
      <td style="padding: 12px; border: 1px solid #ddd; font-weight: bold;">Validation</td>
      <td style="padding: 12px; border: 1px solid #ddd;">0.17758</td>
      <td style="padding: 12px; border: 1px solid #ddd;">0.17846</td>
      <td style="padding: 12px; border: 1px solid #ddd;">0.17758</td>
    </tr>
    <tr>
      <td style="padding: 12px; border: 1px solid #ddd; font-weight: bold; color: #d9534f;">Testing</td>
      <td style="padding: 12px; border: 1px solid #ddd; font-weight: bold;">0.19727</td>
      <td style="padding: 12px; border: 1px solid #ddd; font-weight: bold;">0.19847</td>
      <td style="padding: 12px; border: 1px solid #ddd; font-weight: bold;">0.19727</td>
    </tr>
  </tbody>
</table>

<h3 style="color: #2e64fe; border-bottom: 2px solid #2e64fe; padding-bottom: 5px;">📉 Mean Absolute Error (MAE) Comparison</h3>

<table style="width:100%; border-collapse: collapse; margin: 20px 0; font-family: Arial, sans-serif; box-shadow: 0 4px 8px 0 rgba(0,0,0,0.1);">
  <thead>
    <tr style="background-color: #f2f2f2; text-align: left;">
      <th style="padding: 12px; border: 1px solid #ddd;">Model Type</th>
      <th style="padding: 12px; border: 1px solid #ddd;">Testing MAE Value</th>
      <th style="padding: 12px; border: 1px solid #ddd;">Performance Status</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="padding: 12px; border: 1px solid #ddd; font-weight: bold;">Normal Equation (Manual)</td>
      <td style="padding: 12px; border: 1px solid #ddd;">0.21347</td>
      <td style="padding: 12px; border: 1px solid #ddd; color: #5cb85c; font-weight: bold;">Best (Baseline)</td>
    </tr>
    <tr style="background-color: #f9f9f9;">
      <td style="padding: 12px; border: 1px solid #ddd; font-weight: bold;">Gradient Descent (Manual)</td>
      <td style="padding: 12px; border: 1px solid #ddd;">0.22358</td>
      <td style="padding: 12px; border: 1px solid #ddd;">Very Close</td>
    </tr>
    <tr>
      <td style="padding: 12px; border: 1px solid #ddd; font-weight: bold;">Scikit-Learn Implementation</td>
      <td style="padding: 12px; border: 1px solid #ddd;">0.22197</td>
      <td style="padding: 12px; border: 1px solid #ddd; color: #2196F3;">Optimal</td>
    </tr>
  </tbody>
</table>

<h3 style="color: #2e64fe; border-bottom: 2px solid #2e64fe; padding-bottom: 5px;">📝 Simple Analysis of Results</h3>

<div style="background-color: #f8f9fa; padding: 20px; border-radius: 10px; border: 1px solid #dee2e6;">

<p style="font-size: 16px; line-height: 1.6;">
<strong style="color: #28a745;"> No Overfitting:</strong><br>
 the <b>Testing MSE</b> is very close to the <b>Training MSE</b>. This is a great sign because it means the model is stable and can predict house prices it hasn't seen before, rather than just memorizing the training data.
</p>

<hr style="border: 0; border-top: 1px solid #ccc; margin: 15px 0;">

<p style="font-size: 16px; line-height: 1.6;">
<strong style="color: #17a2b8;"> Methods Comparison:</strong><br>
Both the <b>Normal Equation</b> and <b>Gradient Descent</b> reached almost the same result. This shows that The Gradient Descent implementation worked good and reached the best possible solution.
</p>

<hr style="border: 0; border-top: 1px solid #ccc; margin: 15px 0;">

<p style="font-size: 16px; line-height: 1.6;">
<strong style="color: #6f42c1;"> Code Validation:</strong><br>
<b>from-scratch</b> results are identical to the <b>Scikit-Learn</b> outputs. This gives indication that the manual calculations and NumPy code are correct.
</p>

</div>